### PR TITLE
Improve reliability when communicating with external systems

### PR DESCRIPTION
**Pull request by jsenko**
Any ideas where else to add timeouts or retries?

### CODE REVIEW

File: app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java:134]: I wonder if we should instead map this to HTTP_FORBIDDEN.  I think returning a 404 if the registry tenant is not found has the side effect of leaking information.  This can be used by an attacker to scan for valid tenant IDs.

WDYT?

File: app/src/main/java/io/apicurio/registry/services/auth/CustomAuthenticationMechanism.java
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/services/auth/CustomAuthenticationMechanism.java:123]: Is this where the Retry should happen?  Or should we be more targeted by wrapping the interactions with external systems in CDI beans so we can put the fault tolerance at a more granular location?  Honest question because I'm not sure.

Which lines of code in this method might fail and need to be retried?  It's definitely line 125/130 (OidcAuth).  Is it also 121/126 and 129/134 and 133/138?  It's hard to tell.

Comment (carlesarnal) [reply app/src/main/java/io/apicurio/registry/services/auth/CustomAuthenticationMechanism.java:123]: Agreed, I think we should extract some parts of the code here so we can retry at a more granular level. There are some situations in this method that are worth a retry () but others, IMO, don't.

File: app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java
Comment (carlesarnal) [inline app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java:77]: We don't want this, probably.

**Comment by carlesarnal** [comment]
@jsenko any updates on this? I think we should add the retries relatively soon to our code.

File: app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java
Comment (EricWittmann) [reply app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java:134]: Does the PR from @carlesarnal impact this?  I suspect so.

File: app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java
Comment (EricWittmann) [reply app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java:135]: @carlesarnal wdyt?

File: app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java
Comment (carlesarnal) [reply app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java:134]: Yes, it does.

File: app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java
Comment (carlesarnal) [reply app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java:135]: `RegistryTenantNotFoundException` is thrown by the client, and then is wrapped with TenantNotFound. Actually, that makes me think that the first one should not be present on this mapper.

**Comment by jsenko** [comment]
hi @carlesarnal I reverted changes to the `CustomAuthenticationMechanism`. I'd like to leave the retries in that area to you, as you will know more. I've fixed the small issues from the other comments, so I think we can merge this if tests pass.

**Comment by carlesarnal** [comment]
> hi @carlesarnal I reverted changes to the `CustomAuthenticationMechanism`. I'd like to leave the retries in that area to you, as you will know more. I've fixed the small issues from the other comments, so I think we can merge this if tests pass.

Ok, yes, I will probably refactor that area to break it into smaller pieces. I'll continue once we merge this.

**Review by carlesarnal — APPROVED** [review]
LGTM

**Pull request merged**
